README.md
# nRF905 Arduino API
nRF905 Arduino API. Compatible with ESP8266 and ESP32 boards, tested on *NodeMCU 1.0 (ESP-12E module)* and *DOIT ESP32 Dev Module*.

## Compiling / Installing
1. Open the [nRF905API.ino](https://github.com/eelcohn/nRF905-API/blob/master/src/nRF905API/nRF905API.ino) file in Arduino IDE
2. Open [config.h](https://github.com/eelcohn/nRF905-API/blob/master/src/nRF905API/config.h) and edit the following entries:
    * `WIFI_SSID`: Enter the name of your WiFi network
    * `WIFI_PASSWORD`: Enter the password of your WiFi network
3. Go to Sketch and select Upload

## Connecting the nRF905 module to your ESP8266 or ESP32 module

#### ESP8266 NodeMCU
![NodeMCU pinout](https://github.com/eelcohn/nRF905-API/blob/master/images/nodemcu_pins.png)
| nRF905 pin | ESP8266 pin | ESP8266 GPIO |
|:----------:|:-----------:|:------------:|
|     Vcc    |     3.3V    |    3.3V      |
|     Gnd    |      Gnd    |     Gnd      |
|     AM     |      -      |      -       |
|     CD     |      -      |      -       |
|     CE     |      D2     |    GPIO 4    |
|     DR     |      D1     |    GPIO 5    |
|     PWR    |      D3     |    GPIO 0    |
|    TX_EN   |      D0     |    GPIO 16   |
|    MOSI    | D7 / HMOSI  |    GPIO 13   |
|    MISO    | D6 / HMISO  |    GPIO 12   |
|     CLK    | D5 / HCLK   |    GPIO 14   |
|     CS     | D8 / HCS    |    GPIO 15   |

The NodeMCU doens't have enough GPIO pins for CD and AM to connect to.

#### ESP32
![ESP32 DevKit 30-pin version pinout](https://github.com/eelcohn/nRF905-API/blob/master/images/ESP32-DOIT-DEVKIT-V1-Board-Pinout-30-GPIOs.png)
![ESP32 DevKit 36-pin version pinout](https://github.com/eelcohn/nRF905-API/blob/master/images/ESP32-DOIT-DEVKIT-V1-Board-Pinout-36-GPIOs.jpg)
| nRF905 pin | ESP32 pin | ESP32 GPIO |
|:----------:|:---------:|:----------:|
|     Vcc    |    3.3V   |    3.3V    |
|     Gnd    |    Gnd    |    Gnd     |
|     AM     |    D32    |  GPIO 32   |
|     CD     |    D33    |  GPIO 33   |
|     CE     |    D27    |  GPIO 27   |
|     DR     |    D35    |  GPIO 35   |
|     PWR    |    D26    |  GPIO 26   |
|    TX_EN   |    D25    |  GPIO 25   |
|    MOSI    |    D13    |  GPIO 13   |
|    MISO    |    D12    |  GPIO 12   |
|     CLK    |    D14    |  GPIO 14   |
|     CS     |    D15    |  GPIO 15   |

## Usage

### /api/v1/config.json
Configure the nRF905.

Example:<br>
`http://192.168.x.y/api/v1/config.json?frequency=868400000&crc=16&txpower=10&rxpower=normal&rxaddr=a55a5aa5&txaddr=a55a5aa5&rxaddrwidth=4&txaddrwidth=4&rxpayloadwidth=16&txpayloadwidth=16`

|  Parameter     | Required | Description |
|:--------------:|:--------:|:------------|
|  `frequency`   |    No    | Set the RF frequency |
|     `crc`      |    No    | `off`: disable CRC<br>`8`: enable 8-bit CRC<br>`16`: enable 16-bt CRC |
|   `txpower`    |    No    | Set the RF transmit power<br>`-10`: -10dBm<br>`-2`: -2dBm<br>`6`: 6dBm<br>`10`: 10dBm |
|   `rxpower`    |    No    | Set the RF receive power<br>`normal`: normal receive mode<br>`low`: low power receive mode |
| `txaddrwidth`  |    No    | Set the number of Tx address bytes<br>`1`: 1 byte<br>`4`: 4 bytes |
| `rxaddrwidth`  |    No    | Set the number of Rx address bytes<br>`1`: 1 byte<br>`4`: 4 bytes |
|`txpayloadwidth`|    No    | Set the number of Tx payload bytes<br>`1...32`: Payload is 1...32 bytes |
|`rxpayloadwidth`|    No    | Set the number of Rx payload bytes<br>`1...32`: Payload is 1...32 bytes |
|    `txaddr`    |    No    | Set the Tx address<br>`0x12345678`: Set Tx address to 0x12345678 |
|    `rxaddr`    |    No    | Set the Rx address<br>`0x12345678`: Set Rx address to 0x12345678 |
|    `nvram`     |    No    | Store the nRF905 configuration in NVRAM<br>`true`: Store nRF905 config in NVRAM<br>`false`: Do not store nRF905 config in NVRAM |

### /api/v1/receive.json
This will return a list of received data frames by the nRF905. This API call takes no parameters.

### /api/v1/status.json
This will show a status page with details about the hardware. This API call takes no parameters.

Example:<br>
`https://192.168.x.y/api/v1/status.json`

### /api/v1/send.json
This will transmit data. After the data has been transmitted, the API will wait for data to be received (a reply to the transmitted data). By default it will wait for 2 seconds, but you can change this by adding the `timeout` parameter.

Example:<br>
`https://192.168.x.y/api/v1/send.json?addr=aabbccdd&payload=0400035efa0c04a55a5aa50000000000&timeout=10`

| Parameter | Required | Description |
|:---------:|:--------:|:------------|
| `txaddr`  |   Yes    | Address to send the data to |
| `payload` |   Yes    | The data to be sent |
| `timeout` |    No    | Timeout for reading data after the transmission is completed (in seconds)<br>Default: 2 seconds |

### /api/v1/systemconfig.json
Configure the system board.

Example:<br>
`https://192.168.x.y/api/v1/systemconfig.json?reset=true`

|  Parameter     | Required | Description |
|:--------------:|:--------:|:------------|
|    `reset `    |    No    | Reset the system<br>`true`: Reset system |

## Zehnder API calls

### /api/v1/zehnder/link.json

Link the nRF905-API to a Zehnder device. This API call takes no parameters.

### /api/v1/zehnder/config.json

Configure the Zehnder parameters.

Example:<br>
`https://192.168.x.y/api/v1/zehnder/config.json?network=1a2b3c4d&main_unit_id=91&device_id=5d`

|  Parameter     | Required | Description |
|:--------------:|:--------:|:------------|
|   `network`    |   Yes    | Network ID of your Zehnder network |
| `main_unit_id` |   Yes    | ID of your Zehnder main unit |
|  `device_id`   |   Yes    | ID of your nRF905-API |

### /api/v1/zehnder/setpower.json

Set the power level of your Zehnder device.

Example:<br>
`https://192.168.x.y/api/v1/zehnder/setpower.json?power=high`

|  Parameter     | Required | Description |
|:--------------:|:--------:|:------------|
|    `power`     |   Yes    | Power level:<br>`low`: Low power<br>`medium`: Medium power<br>`high`: High power |

### /api/v1/zehnder/settimer.json

Set the timer of your Zehnder device.

Example:<br>
`https://192.168.x.y/api/v1/zehnder/settimer.json?power=high&minutes=5`

|  Parameter     | Required | Description |
|:--------------:|:--------:|:------------|
|    `power`     |   Yes    | Power level:<br>`low`: Low power<br>`medium`: Medium power<br>`high`: High power |
|   `minutes`    |   Yes    | Number of minutes |



CHANGELOG
v0.3.2
* Fixed autoconfiguration for endpoint /api/v1/zehnder/config.json
* Fixed autoconfiguration for endpoint /api/v1/zehnder/link.json

v0.3.1
* Added endpoint /api/v1/zehnder/config.json
* Added endpoint /api/v1/zehnder/link.json
* Added endpoint /api/v1/zehnder/setpower.json
* Added endpoint /api/v1/zehnder/settimer.json
* Fixed stability issues

v0.3.0
* Renamed endpoints to .json
* Added /api/v1/systemconfig.json endpoint
* Set nRF905 to idle mode when writing config file
* Updated switching nRF905 modes when transmitting data
* Added storing the nRF905 config settings in NVRAM/EEPROM
* Fixed voltage measurement for ESP8266 boards
* Fixed stability issues

v0.2.1
* First public beta release

v0.2.0
* Unreleased test version

v0.1.1
* Unreleased test version

v0.1.0
* First test version



src/nRF905API/esp32.h
#ifndef __NRF905_ESP32_H__
#define __NRF905_ESP32_H__

#ifdef ARDUINO_ARCH_ESP32

#include <SPI.h>	// SPIClass

#if NRF905API_HTTPS == 1
#include <WebServerSecure.h>
extern WebServerSecure server;
#else
#include <WebServer.h>
extern WebServer server;
#endif
#if NRF905API_MDNS == 1
#include <ESPmDNS.h>
#endif
#if NRF905API_SSDP == 1
#include <SSDP.h>
#endif

// ESP32 devboard pin definitions
#ifndef PIN_AM
#define PIN_AM   32	// nRF905 AM pin (Address Match)
#endif
#ifndef PIN_CD
#define PIN_CD   33	// nRF905 CD pin (Carrier detect)
#endif
#ifndef PIN_CE
#define PIN_CE   27	// nRF905 CS pin (Enable pin)
#endif
#ifndef PIN_DR
#define PIN_DR   35	// nRF905 DR pin (Data Ready)
#endif
#ifndef PIN_PWR
#define PIN_PWR  26	// nRF905 PWR pin (Power mode pin)
#endif
#ifndef PIN_TXEN
#define PIN_TXEN 25	// nRF905 TX_EN pin (Transmit enable pin)
#endif
#ifndef PIN_MOSI
#define PIN_MOSI 13	// nRF905 SPI MOSI pin
#endif
#ifndef PIN_MISO
#define PIN_MISO 12	// nRF905 SPI MISO pin
#endif
#ifndef PIN_SPICLK
#define PIN_SPICLK 14	// nRF905 SPI clock pin
#endif
#ifndef PIN_SPICS
#define PIN_SPICS 15	// nRF905 SPI CS pin
#endif

class Board {
	public:
		const String	get_arch(void);
		void		reset(void);
		const String	restartReason(void);
		const char *	getDateTime(void);
		const uint32_t	getCPUFreqMhz(void);
		const char *	getSdkVersion(void);
		void		setADCtoVccMode(void);
		const uint16_t	get_vcc(void);
		const uint32_t	get_cpu_id(void);
		const uint32_t	get_flash_chip_id(void);
		const uint32_t	getFlashChipSpeed(void);
		const bool	checkFlashCRC(void);
		const char *	getFlashMode(void);
		const uint32_t	get_flash_chip_real_size(void);
		const uint32_t	getFlashChipSdkSize(void);
		const uint32_t	getSketchSize(void);
		const uint32_t	getFreeSketchSpace(void);
		const String	getSketchMD5(void);
		const uint32_t	getFreeHeap(void);
		const uint32_t	getHeapFragmentation(void);
		const uint32_t	getHeapMaxFreeBlockSize(void);
		const String	get_core_version(void);
		const uint32_t	get_core_revision(void);
		String		get_hostname(void);
		void		set_hostname(const char *hostname);
		String		get_localIPv6(void);
		bool		enable_IPv6(void);
		void		setPinMode(const uint8_t pin, const uint8_t mode);
		void		writePin(const uint8_t pin, const uint8_t value);
		uint8_t	readPin(const uint8_t pin);
		void		onBoardLED(const bool value);
		bool		SPIBegin(const uint8_t mosi, const uint8_t miso, const uint8_t clk, const uint8_t cs);
		void		SPIEnd(void);
		void		SPISetChipSelectPolarity (const uint8_t cs, const uint8_t active);
		void		SPISetBitOrder(const uint8_t order);
		void		SPISetDataMode(const uint8_t mode);
		void		SPISetFrequency(const uint32_t frequency);
		uint8_t	SPITransfer(const uint8_t out);
		void		SPITransfern(uint8_t * buffer, const size_t size);
		void		ReadNVRAM(uint8_t * buffer, const uint32_t offset, const size_t size);
		void		WriteNVRAM(const uint8_t * buffer, const uint32_t offset, const size_t size);

	private:
		uint32_t	_spi_frequency;
		uint8_t	_spi_bitorder;
		uint8_t	_spi_datamode;
		uint8_t	_spi_cs;
		uint8_t	_spi_cs_active;
		SPIClass *	spi = NULL;
};

#endif

#endif

